search/includes/classes/class-search.php

```diff
@@ -572,19 +572,48 @@ public function filter__ep_do_intercept_request( $request, $query, $args, $failu
 			}
 		} else {
 			// Record engine time (have to parse JSON to get it)
-			$response_body = wp_remote_retrieve_body( $request );
-			$response = json_decode( $response_body, true );
+			$response_body_json = wp_remote_retrieve_body( $response );
+			$response_body = json_decode( $response_body_json, true );
 
-			if ( $response && isset( $response['took'] ) && is_int( $response['took'] ) ) {
-				$statsd->timing( $statsd_prefix . '.engine', $response['took'] );
-				$statsd->timing( $statsd_per_site_prefix . '.engine', $response['took'] );
+			if ( $response_body && isset( $response_body['took'] ) && is_int( $response_body['took'] ) ) {
+				$statsd->timing( $statsd_prefix . '.engine', $response_body['took'] );
+				$statsd->timing( $statsd_per_site_prefix . '.engine', $response_body['took'] );
 			}
 
 			$statsd->timing( $statsd_prefix . '.total', $duration );
 			$statsd->timing( $statsd_per_site_prefix . '.total', $duration );
+
+			$response_code = (int) wp_remote_retrieve_response_code( $response );
+
+			// Check for an error HTTP code and log the Elasticsearch error
+			if ( $response_code >= 400 ) {
+				$response_error = $response_body['error'];
+				$error_message = $response_error['reason'] ?? 'Unknown Elasticsearch query error';
+				\Automattic\VIP\Logstash\log2logstash( array(
```

Comment:
For this and the `Warning` header below, should we also call `trigger_error` with `E_USER_WARNING` to give customers visibility into these errors as well? (With enough volume, it would get surfaced in New Relic.)

Reply:
Oh, it certainly makes sense! One day I'll need to investigate what are all the paths for error logging, as this is still a bit obscure for me :sweat_smile:

Reply:
For the moment I've only added this to the Warning header condition, as in that case I can trust more the message. For the 400/500 error responses, there's more diversity, and often the message just becomes "all shards failed". It means that all of them failed to execute the query because it was wrong, but sending an "all shards failed" warning to New Relic can be very misleading.

I'll monitor what we receive in logstash (with the full detail of the error) and will decide what's the best way to trigger a useful warning here.